packages/plugin-workspace-tools/sources/commands/foreach.ts

```diff
@@ -116,9 +126,14 @@ export default class WorkspacesForeachCommand extends BaseCommand {
       ? project.topLevelWorkspace
       : cwdWorkspace!;
 
+    const fromPredicate = (workspace: Workspace) => micromatch.isMatch(structUtils.stringifyIdent(workspace.locator), this.from);
+    const fromCandidates: Array<Workspace> = this.from.length
+      ? [rootWorkspace, ...rootWorkspace.getRecursiveWorkspaceChildren()].filter(fromPredicate)
+      : [rootWorkspace];
+
     const candidates = this.recursive
```

Comment:
This will contain duplicates, but that's taken care of when actually iterating further down.

Reply:
noahnu:feat_2834